packages/artifact/src/internal/download-http-client.ts

```diff
@@ -263,26 +263,47 @@ export class DownloadHttpClient {
       if (isGzip) {
         const gunzip = zlib.createGunzip()
         response.message
+          .on('error', error => {
```

Comment:
The existing `on('error', ...)` callback below isn't called when gunzip fails, or response.message produces an error. `pipe` doesn't propagate errors across streams, so you have to monitor each.

Without the new blocks the error is unhandled, and the new test times out.

I'm closing the downstream stream (hah!) on error. From https://nodejs.org/dist/latest-v14.x/docs/api/stream.html#stream_readable_pipe_destination_options:

> One important caveat is that if the `Readable` stream emits an error during processing, the `Writable` destination is not closed automatically. If an error occurs, it will be necessary to manually close each stream in order to prevent memory leaks.

Reply:
I am a little bit unclear about when to retry here.  If the download failed because the socket is opened too long (for downloading large files), would retry help?

Reply:
Good question. In short, I'm hoping retries will help. 😄 

In the case of smaller files, there are likely transient requests that stall out or timeout that will be quicker on retry.

For larger files (https://github.com/actions/download-artifact/issues/55), is there a technical reason we'd be able to upload large files we can't download? Do we only chunk on upload? Should we add a max size check to upload-artifact if download-artifact is unlikely to succeed?

One option to add retries to hopefully improve the success rate for smaller files, and then revisit this code if customers still report issues with larger files.

Reply:
We do chunked upload but I don't believe we do chunked download.  

> and then revisit this code if customers still report issues with larger files.

This is what I am afraid of, since in https://github.com/actions/download-artifact/issues/55 the description says there is a 100% failure rate, I am afraid the retry will be in vain, and we will incur more minutes per build from those retries.